R/verb-select.R

```diff
@@ -83,26 +83,53 @@ rename_with.tbl_lazy <- function(.data, .fn, .cols = everything(), ...) {
 #' @inheritParams dplyr::relocate
 #' @export
 relocate.tbl_lazy <- function(.data, ..., .before = NULL, .after = NULL) {
+  # Hack: We want to use `dplyr::relocate.data.frame()` instead of reimplementing it.
+  # Because `relocate()` can rename columns we use an attribute to store the
+  # original column position.
+  sim_data <- simulate_vars(.data)
+  for (i in seq_along(sim_data)) {
+    attr(sim_data[[i]], "dbplyr_org_pos") <- i
+  }
+
   new_vars <- dplyr::relocate(
-    simulate_vars(.data),
+    sim_data,
     ...,
     .before = {{.before}},
     .after = {{.after}}
   )
 
-  .data$lazy_query <- add_select(.data, syms(unlist(new_vars)))
+  old_vars <- colnames(sim_data)
+  vars_mapping <- purrr::map_chr(
+    new_vars,
+    ~ old_vars[[attr(.x, "dbplyr_org_pos")]]
+  )
+
+  .data$lazy_query <- add_select(.data, syms(vars_mapping))
   .data
 }
 
-simulate_vars <- function(x, drop_groups = FALSE) {
+#' Simulate variables to use in tidyselect
+#'
+#' @param x A lazy table
+#' @param drop_groups Should groups be dropped?
+#'
+#' @return A 0 row tibble with the same columns names, and, if possible, types, as `x`.
+#'
+#' @export
+#' @keywords internal
+simulate_vars <- function (x, drop_groups = FALSE) {
+  UseMethod("simulate_vars")
+}
+
+#' @export
+simulate_vars.tbl_lazy <- function(x, drop_groups = FALSE) {
```

Comment:
I'd suggested adding something like:

```R
simulate_vars_is_typed <- function(x) UseMethod("simulate_vars_is_typed")
simulate_vars_is_typed.default <- function(x) FALSE
```

Then sparklyr can implement that generic at the same time, and we'll be in a good place for the next release of tidyselect. If you do that, I think it's fine to merge this PR.